In [11]:
# Importing all library required for analysis
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
%matplotlib inline
sns.set()
pd.options.display.max_columns = None
pd.options.display.max_rows = None

### Descriptive Analysis

In [2]:
# Reading the csv file
df = pd.read_csv('tripdata_upto_oct_2021_cleaned.csv')

Backing off send_request(...) for 268.1s (requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.segment.io', port=443): Max retries exceeded with url: /v1/batch (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002584B6F99D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')))
c:\users\detective\miniconda3\lib\site-packages\IPython\core\interactiveshell.py:3457: DtypeWarning: Columns (7,9) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)
Giving up send_request(...) after 11 tries (requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.segment.io', port=443): Max retries exceeded with url: /v1/batch (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002584B6F95E0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')))
error uploading: HTTPSConnectionPool(host='api.segment.io', port

In [15]:
# Creating a backup copy
data = df.copy()

In [18]:
# Drop unnecessary column
data = data.drop(['Unnamed: 0', 'Unnamed: 0.1'],  axis = 1)

In [19]:
data.head()

ride_id  rideable_type           started_at             ended_at  \
0  BD0A6FF6FFF9B921  electric_bike  2020-11-01 13:36:00  2020-11-01 13:45:40   
1  96A7A7A4BDE4F82D  electric_bike  2020-11-01 10:03:26  2020-11-01 10:14:45   
2  C61526D06582BDC5  electric_bike  2020-11-01 00:34:05  2020-11-01 01:03:06   
3  E533E89C32080B9E  electric_bike  2020-11-01 00:45:16  2020-11-01 00:54:31   
4  1C9F4EF18C168C60  electric_bike  2020-11-01 15:43:25  2020-11-01 16:16:52   

          start_station_name start_station_id          end_station_name  \
0      Dearborn St & Erie St            110.0    St. Clair St & Erie St   
1  Franklin St & Illinois St            672.0  Noble St & Milwaukee Ave   
2  Lake Shore Dr & Monroe St             76.0      Federal St & Polk St   
3   Leavitt St & Chicago Ave            659.0   Stave St & Armitage Ave   
4        Buckingham Fountain              2.0       Buckingham Fountain   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  \
0          211.0  41.894177 -87.629127  41.894434 -87.623379        casual   
1           29.0  41.890959 -87.635343  41.900675 -87.662480        casual   
2           41.0  41.880983 -87.616754  41.872054 -87.629550        casual   
3          185.0  41.895499 -87.682013  41.917744 -87.691392        casual   
4            2.0  41.876497 -87.620358  41.876448 -87.620338        casual   

         ride_time day_of_week  
0  0 days 00:09:40      Sunday  
1  0 days 00:11:19      Sunday  
2  0 days 00:29:01      Sunday  
3  0 days 00:09:15      Sunday  
4  0 days 00:33:27      Sunday

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4491323 entries, 0 to 4491322
Data columns (total 15 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
 13  ride_time           object 
 14  day_of_week         object 
dtypes: float64(4), object(11)
memory usage: 514.0+ MB


In [14]:
# Random sampling without replacement (due to low computation power)
# Population - 44,91,323
# Confidence Interval - 99%
# Margin of error - 0.15%
# Sample size - 6,35,029

#new_data = data.sample(n = 635029)

In [15]:
# Write sample dataset to a new csv file
#new_data.to_csv('sample_dataset.csv')

In [22]:
# Filter latittude and longitudes (rows with 0.0 values)
#filt = (new_data.iloc[:,8:12].values != 0.0)
#new_data = new_data.loc[filt]

In [ ]:
# Calculating ride distance (miles)
#from geopy.distance import geodesic

#new_data['ride_distance'] = new_data.apply(lambda x: geodesic((x['start_lat'], x['start_lng']), (x['end_lat'], x['end_lng'])).m, axis=1)

In [22]:
# Write sample dataset to a new csv file
#new_data.to_csv('sample_dataset_v1.csv')

In [25]:
# Percentage of members and casual users
data['member_casual'].value_counts(normalize = True)*100

member    54.788222
casual    45.211778
Name: member_casual, dtype: float64

In [26]:
#Percentage of bike type used for ride
data['rideable_type'].value_counts(normalize = True)*100

classic_bike     68.094813
electric_bike    21.602120
docked_bike      10.303067
Name: rideable_type, dtype: float64

In [27]:
#Ride Started Stations (percentages)
data['start_station_name'].value_counts(normalize = True)*100

Streeter Dr & Grand Ave                                  1.754472
Michigan Ave & Oak St                                    0.958493
Wells St & Concord Ln                                    0.906236
Millennium Park                                          0.888224
Clark St & Elm St                                        0.869989
Theater on the Lake                                      0.822787
Wells St & Elm St                                        0.783110
Clark St & Lincoln Ave                                   0.709791
Clark St & Armitage Ave                                  0.700840
Wabash Ave & Grand Ave                                   0.684809
Kingsbury St & Kinzie St                                 0.678686
Dearborn St & Erie St                                    0.671183
Wells St & Huron St                                      0.653171
Broadway & Barry Ave                                     0.647938
Indiana Ave & Roosevelt Rd                               0.626207
St. Clair 

In [28]:
#Ride Ended Stations (percentages)
data['end_station_name'].value_counts(normalize = True)*100

Streeter Dr & Grand Ave                                  1.779832
Michigan Ave & Oak St                                    0.974390
Wells St & Concord Ln                                    0.917057
Millennium Park                                          0.915031
Clark St & Elm St                                        0.861862
Theater on the Lake                                      0.834832
Wells St & Elm St                                        0.782331
Clark St & Lincoln Ave                                   0.710503
Wabash Ave & Grand Ave                                   0.702599
Clark St & Armitage Ave                                  0.687370
Dearborn St & Erie St                                    0.682293
Kingsbury St & Kinzie St                                 0.673966
Broadway & Barry Ave                                     0.660251
St. Clair St & Erie St                                   0.649385
Wells St & Huron St                                      0.632375
Lake Shore

In [29]:
#Grouping by member_casual column
customer_grp = data.groupby(['member_casual'])

In [30]:
# Bike type used by members and casual users (%)
customer_grp['rideable_type'].value_counts(normalize = True)*100

member_casual  rideable_type
casual         classic_bike     60.148517
               electric_bike    22.601665
               docked_bike      17.249817
member         classic_bike     74.652174
               electric_bike    20.777286
               docked_bike       4.570540
Name: rideable_type, dtype: float64

In [31]:
# Converting ride time to minutes
data['ride_time_min'] = data['ride_time']/np.timedelta64(1,'m')

In [32]:
#Ride time of Casual Users(minutes)
customer_grp['ride_time_min'].agg(['median', 'mean', 'max', 'min']).loc['casual']

median       16.966667
mean         33.001642
max       55944.150000
min           0.016667
Name: casual, dtype: float64

In [33]:
#Ride time of members(minutes)
customer_grp['ride_time_min'].agg(['median', 'mean', 'max', 'min']).loc['member']

median       9.950000
mean        13.479885
max       1499.933333
min          0.016667
Name: member, dtype: float64

In [34]:
#Day of ride - Casual Users (%)
customer_grp['day_of_week'].value_counts(normalize = True).loc['casual']*100

day_of_week
Saturday     23.049512
Sunday       19.877751
Friday       14.148676
Monday       11.141644
Thursday     10.902208
Wednesday    10.503657
Tuesday      10.376552
Name: day_of_week, dtype: float64

In [35]:
#Day of ride - Members (%)
customer_grp['day_of_week'].value_counts(normalize = True).loc['member']*100

day_of_week
Wednesday    15.385156
Tuesday      14.929679
Thursday     14.643583
Friday       14.479282
Saturday     14.392478
Monday       13.547073
Sunday       12.622749
Name: day_of_week, dtype: float64

In [36]:
new_df = pd.DataFrame(data, columns = ['rideable_type', 'start_station_name', 'end_station_name','member_casual', 'day_of_week', 'ride_time_min'])

In [37]:
new_df.head()

rideable_type         start_station_name          end_station_name  \
0  electric_bike      Dearborn St & Erie St    St. Clair St & Erie St   
1  electric_bike  Franklin St & Illinois St  Noble St & Milwaukee Ave   
2  electric_bike  Lake Shore Dr & Monroe St      Federal St & Polk St   
3  electric_bike   Leavitt St & Chicago Ave   Stave St & Armitage Ave   
4  electric_bike        Buckingham Fountain       Buckingham Fountain   

  member_casual day_of_week  ride_time_min  
0        casual      Sunday       9.666667  
1        casual      Sunday      11.316667  
2        casual      Sunday      29.016667  
3        casual      Sunday       9.250000  
4        casual      Sunday      33.450000

In [38]:
#Data visualization
import sweetviz as sv

my_report = sv.compare_intra(new_df, new_df["member_casual"] == "casual", ["member", "casual"])

my_report.show_html()

#Data visualization is provided as html file (SWEETVIZ_REPORT.html)

                                             |                                             | [  0%]   00:03 ->…

Report SWEETVIZ_REPORT.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


### Observations:
- There is more members (55%) than casual users (45%)
- Classical Bike (68%) is the commonly used ride by both casual users and members
- Casual users ride longer than members
- Casual riders ride during weekends (12.5% Higher than members)

#### Suggestions for converting casual users to members
* Offers on weekend rides
* Offers on long distance rides
* Offers displayed in the primary start and end stations